src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -0,0 +1,657 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Reflection.Metadata.Ecma335;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.ProjectSystem;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Workspaces.ProjectSystem
+{
+    internal sealed class ProjectSystemProjectFactory
+    {
+        /// <summary>
+        /// The main gate to synchronize updates to this solution.
+        /// </summary>
+        /// <remarks>
+        /// See the Readme.md in this directory for further comments about threading in this area.
+        /// </remarks>
+        // TODO: we should be able to get rid of this gate in favor of just calling the various workspace methods that acquire the Workspace's
+        // serialization lock and then allow us to update our own state under that lock.
+        private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
+
+        public Workspace Workspace { get; }
+        public IAsynchronousOperationListener WorkspaceListener { get; }
+        public IFileChangeWatcher FileChangeWatcher { get; }
+        public FileWatchedPortableExecutableReferenceFactory FileWatchedReferenceFactory { get; }
+
+        private readonly Action<ImmutableArray<string>> _onDocumentsAdded;
+        private readonly Action<Project> _onProjectRemoved;
+
+        /// <summary>
+        /// A set of documents that were added by <see cref="ProjectSystemProject.AddSourceTextContainer"/>, and aren't otherwise
+        /// tracked for opening/closing.
+        /// </summary>
+        public ImmutableHashSet<DocumentId> DocumentsNotFromFiles { get; private set; } = ImmutableHashSet<DocumentId>.Empty;
+
+        /// <remarks>Should be updated with <see cref="ImmutableInterlocked"/>.</remarks>
+        private ImmutableDictionary<ProjectId, string?> _projectToMaxSupportedLangVersionMap = ImmutableDictionary<ProjectId, string?>.Empty;
+
+        /// <remarks>Should be updated with <see cref="ImmutableInterlocked"/>.</remarks>
+        private ImmutableDictionary<ProjectId, string> _projectToDependencyNodeTargetIdentifier = ImmutableDictionary<ProjectId, string>.Empty;
+
+        /// <summary>
+        /// Set by the host if the solution is currently closing; this can be used to optimize some things there.
+        /// </summary>
+        public bool SolutionClosing { get; set; }
+
+        /// <summary>
+        /// The current path to the solution. Currently this is only used to update the solution path when the first project is added -- we don't have a concept
+        /// of the solution path changing in the middle while a bunch of projects are loaded.
+        /// </summary>
+        public string? SolutionPath { get; set; }
```

Comment:
maybe say that if we get a new solution, we'd have a new factory or something?  it's unclear the relationship here.